design-proposals/deploy-experience-improvement.md

```diff
@@ -0,0 +1,240 @@
+
+- The cloud installer and infrastructure provisioning are implemented as **large monolithic shell scripts**.
+  This not only makes them difficult to maintain but also renders unit testing nearly impossible.
+- The installer is **only tested late in the hardware integration pipeline (HIP)**, which delays feedback and makes bugs harder to trace and resolve.
+- The on-prem installer was developed in parallel but shares little code or structure with the cloud installer.
+  This results in **inconsistent behaviors and duplicated logic between cloud and on-prem** deployments.
+- There is **no clear boundary between infrastructure provisioning and orchestrator** setup,
+  making it difficult to port components to another cloud provider or isolate issues during upgrade and testing.
+- **Upgrade support was added as a second thought** and lacks proper design.
+- **Error handling is poor**; raw error logs are surfaced directly to users with no actionable remediation,
+  and failures require rerunning entire stages without guidance.
+
+This proposal aims to significantly improve the deployment experience of EMF across multiple environments (AWS, Azure, and on-prem).
+The new installer will prioritize user experience by offering a streamlined, zero-touch installation process after configuration,
+along with clear error handling and actionable feedback.
+It will also increase cloud portability through clear infrastructure abstraction and support for Azure.
+Finally, by replacing monolithic shell scripts with modular Go components and adding test coverage, we will enable faster iteration and more frequent releases.
+
+## Proposal
+
+### Scope of work
+
+- A **unified installer** that supports AWS, Azure, and on-prem targets.
+- A **text user interface (TUI) interactive configuration builder** that guides users through required inputs,
+  with the ability to preload values from environment variables or prior installations.
+  It should minimize user input. For example, scale profiles for infra and orchestrator can be automatically applied according to user-specified target scale.
+- A **well-defined abstraction between infrastructure and orchestrator** logic that enables independent testing and upgrading,
+  as well as the ability to plug in new cloud providers via Go modules.
+- **Wrapped and actionable error messages**. Raw logs should be saved to files, and restarts should be possible from the point of failure.
+- Diff previews should be available during upgrade flows, showing schema migrations or configuration changes.
+- Every module should be **toggled independently** and have minimal external dependency.
+- Installer should support **orchestrator CLI integration** (e.g. `cli deploy aws`) and parallel execution of non-dependent tasks.
+- On-prem installation will not require a separate admin machine.
+- Be compatible with upcoming Azure implementation and ongoing replacement of kind with on-prem in Coder
```

Comment:
:(. Uff I was hoping for kind

Reply:
There are several benefits of focusing on the same deployment method that customers use.

We acknowledge that on-prem infra is not as lightweight as it can be, and we are committed to proposing and implementing a weight-loss program.